torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -482,6 +488,43 @@ ProcessGroupNCCL::ProcessGroupNCCL(
     asyncErrorHandling_ = false;
   }
 
+  // Perform health check by initializing dummy communicators and destroying
+  // them. This will help indicate any NCCL-related issues prior to the first
+  // collective.
+  // Run it in a separate thread and wait on CV to handle timeouts, since
+  // majority of getNCCLComm failures are hangs.
+  std::mutex healthCheckMutex;
+  std::condition_variable healthCheckCv;
+  bool healthCheckSuccess = false;
+  auto t = std::thread([&healthCheckMutex, &healthCheckCv, &healthCheckSuccess, this]() {
+    std::vector<at::Device> rankDevice = {getDeviceForRank(rank_)};
```

Comment:
We probably need some signaling mechanism from this thread to the main thread. If this thread dies due to an exception, the main thread should know about it and it should be reported. Probably need to catch the exception and store it in an exception_ptr that can be queried by the main thread.

Reply:
I mentioned this on 494-495, but the vast majority of errors we expect here (and seen in internal training workloads) is hangs. As a result we don't get an exception we can signal the main thread with. 

To deal with the hang, I added a wait on `healthCheckCv` so that if this thread doesn't signal in time (whether it died due to exception or hang), we get a timeout on the main thread and indicate the health check failed. 

Seems like this is enough to indicate something went wrong with communicator initialization, but in the cases of exceptions I guess we could wrap all this in a try-catch and populate an `exception_ptr`?